### PR TITLE
[KOGITO-829] Migrating Data Index to Deployment

### PR DESCRIPTION
**Pull request by ricardozanini**
See:
https://issues.redhat.com/browse/KOGITO-829

In this PR we changed the way the Data Index is deployed. Instead of using `Statefulset`, we're relying on `Deployment`, much like what Jobs Service does.

TODO:

- [x] Create an Image Stream when on OpenShift
- [x] Run Smoke tests with the new approach :heart: 

Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.MD#sending-a-pull-request)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

### CODE REVIEW

File: deploy/crds/app.kiegroup.org_kogitodataindices_crd.yaml [hunk 1]
```diff
@@ -171,25 +171,33 @@ spec:
               description: Status of the Data Index Service Deployment created and
                 managed by it
               properties:
+                availableReplicas:
+                  description: Total number of available pods (ready for at least
+                    minReadySeconds) targeted by this deployment.
+                  format: int32
+                  type: integer
                 collisionCount:
```
Comment (sutaakar) [inline deploy/crds/app.kiegroup.org_kogitodataindices_crd.yaml]: Is this field still needed?

Comment (ricardozanini) [reply deploy/crds/app.kiegroup.org_kogitodataindices_crd.yaml]: Those are auto generated by the Deployment Status type, there's not much to do unfortunately.

File: pkg/framework/image_stream.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import "fmt"
+
+const (
+	annotationKeyImageTriggers         = "image.openshift.io/triggers"
+	annotationValueImageTriggersFormat = "[{\"from\":{\"kind\":\"ImageStreamTag\",\"name\":\"%s\"},\"fieldPath\":\"spec.template.spec.containers[?(@.name==\\\"%s\\\")].image\"}]"
+)
+
+// ResolveImageStreamTriggerAnnotation creates a key and value combination for the ImageStream trigger to be linked with a Kubernetes Deployment
+// this way, a Deployment resource can be attached to a ImageStream, like the DeploymentConfigs are.
+// See: https://docs.openshift.com/container-platform/3.11/dev_guide/managing_images.html#image-stream-kubernetes-resources
```
Comment (sutaakar) [inline pkg/framework/image_stream.go]: Is this supported also for OpenShift 4? I haven't found it in doc anywhere for 4.3.

Comment (ricardozanini) [reply pkg/framework/image_stream.go]: Yes, it's supported on OCP 4.x. The Jobs Service was already built on top of it. It's a nice feature to have a Kubernetes resource to talk to an OpenShift one. Thinking ahead, will make sense for us when supporting Kubernetes.

File: pkg/controller/kogitoapp/resource/kogito_imagestream.go [hunk 3]
```diff
@@ -29,32 +28,29 @@ import (
 	imgv1 "github.com/openshift/api/image/v1"
 )
 
-const (
```
Comment (spolti) [inline pkg/controller/kogitoapp/resource/kogito_imagestream.go]: why removing this block of contants and declare it individually?

File: pkg/controller/kogitoapp/resource/kogito_imagestream_test.go [hunk 4]
```diff
@@ -89,14 +89,6 @@ func TestKogitoImageStreamGeneration(t *testing.T) {
 			assert.Equal(t, imgv1.TagReferencePolicy{Type: imgv1.LocalTagReferencePolicy}, item.Spec.Tags[0].ReferencePolicy)
 			assert.Equal(t, "DockerImage", item.Spec.Tags[0].From.Kind)
 			assert.Equal(t, fmt.Sprintf("quay.io/kiegroup/%s:%s", item.Name, version.Version), item.Spec.Tags[0].From.Name)
-
-		case KogitoDataIndexImage:
```
Comment (spolti) [inline pkg/controller/kogitoapp/resource/kogito_imagestream_test.go]: shouldn't we keep this?

File: pkg/controller/kogitodataindex/resource/kafka_test.go [hunk 5]
```diff
@@ -133,37 +133,6 @@ func TestIsKafkaServerURIResolved(t *testing.T) {
 	}
 }
 
-func Test_fromKafkaToStringMap(t *testing.T) {
```
Comment (spolti) [inline pkg/controller/kogitodataindex/resource/kafka_test.go]: shouldn't we keep this test?

File: test/smoke/framework/kogitodataindexservice.go [hunk 6]
```diff
@@ -40,3 +42,29 @@ func DeployKogitoDataIndexService(namespace string, replicas int) error {
 	}
 	return nil
 }
+
+
+// GetKogitoJobsService retrieves the running jobs service
+func GetKogitoDataIndex(namespace string) (*v1alpha1.KogitoDataIndex, error) {
+	service := &v1alpha1.KogitoDataIndex{}
```
Comment (radtriste) [inline test/smoke/framework/kogitodataindexservice.go]: can you change variable's name ?

File: pkg/controller/kogitoapp/resource/kogito_imagestream_test.go (hunk 4 above)
Comment (ricardozanini) [reply pkg/controller/kogitoapp/resource/kogito_imagestream_test.go]: No, the image stream is being created by the Data Index controller.

File: pkg/controller/kogitoapp/resource/kogito_imagestream.go (hunk 3 above)
Comment (ricardozanini) [reply pkg/controller/kogitoapp/resource/kogito_imagestream.go]: Goland lint.

File: pkg/controller/kogitodataindex/resource/kafka_test.go (hunk 5 above)
Comment (ricardozanini) [reply pkg/controller/kogitodataindex/resource/kafka_test.go]: We don't have this function anymore, the variables are being tested in another use case.

File: test/smoke/framework/kogitodataindexservice.go (hunk 6 above)
Comment (ricardozanini) [reply test/smoke/framework/kogitodataindexservice.go]: For sure!

**Comment by ricardozanini** [comment]
Guys, I'm going to merge this since I'll be out for the whole week, would be nice if we catch any bugs in this feature during this week.

Thank you @spolti for your review at the end of the day.

Thanks!

File: pkg/controller/kogitodataindex/resource/resource_comparator.go [hunk 7]
```diff
@@ -32,15 +34,28 @@ func GetComparator() compare.MapComparator {
...
 			WithType(reflect.TypeOf(routev1.Route{})).
 			WithCustomComparator(framework.CreateRouteComparator()).
 			UseDefaultComparator().
-			BuildAsFunc())
+			Build())
+
+	resourceComparator.SetComparator(
+		framework.NewComparatorBuilder().
+			WithType(reflect.TypeOf(appsv1.Deployment{})).
+			UseDefaultComparator().
```
Comment (xiezhang7) [inline pkg/controller/kogitodataindex/resource/resource_comparator.go]: It is not necessary to set a comparator if you want to use the default one.

Comment (ricardozanini) [reply pkg/controller/kogitodataindex/resource/resource_comparator.go]: Yes, the funny thing is that I wrote the API and I forget about it. I might need to do some refactor to it.

**Pull request merged**